src/math/CurveFitting.cc

```diff
@@ -235,64 +212,89 @@ namespace pfs{ namespace drp{ namespace stella{ namespace math{
+                                                        keyWords,
+                                                        args);
       #ifdef __DEBUG_POLYFIT__
         cout << "pfs::drp::stella::math::CurveFitting::PolyFit: PolyFit(D_A1_XArr, D_A1_YArr, I_Degree_In, S_A1_Args_In, ArgV) returned D_A1_Coeffs_Out = " << D_A1_Coeffs_Out << endl;
       #endif
-      ndarray::Array<T, 1, 1> D_A1_YFit;
-      D_A1_YFit = pfs::drp::stella::math::Poly(D_A1_XArr, 
-                                               D_A1_Coeffs_Out,
-                                               T(xRange[0]),
-                                               T(xRange[1]));
-
+      ndarray::Array<T, 1, 1> D_A1_YFit = pfs::drp::stella::math::Poly(D_A1_XArr, 
+                                                                       D_A1_Coeffs_Out,
+                                                                       xRange[0],
+                                                                       xRange[1]);
+      #ifdef __DEBUG_POLYFIT__
+        cout << "pfs::drp::stella::math::CurveFitting::PolyFit: D_A1_YFit = " << D_A1_YFit << endl;
+      #endif
       ndarray::Array<T, 1, 1> D_A1_Temp = ndarray::allocate(D_A1_Y.size());
       for (int pos = 0; pos < D_A1_Y.size(); ++pos)
         D_A1_Temp[pos] = D_A1_Y[pos] - D_A1_YFit[pos];
       Eigen::Array<T, Eigen::Dynamic, 1> tempEArr = D_A1_Temp.asEigen();
       D_A1_Temp.asEigen() = tempEArr.pow(2) / T(D_A1_Y.size());
-//        D_A1_Temp.deep() = D_A1_Temp / D_A1_Y.size();
       double D_SDev = double(sqrt(D_A1_Temp.asEigen().sum()));
 
-      D_A1_PolyRes.deep() = pfs::drp::stella::math::Poly(xNew, 
+      D_A1_PolyRes.deep() = pfs::drp::stella::math::Poly(D_A1_X_In,
                                                          D_A1_Coeffs_Out,
-                                                         T(xRange[0]),
-                                                         T(xRange[1]));
+                                                         xRange[0],
+                                                         xRange[1]);
       double D_Dev;
       D_A1_X.resize(0);
       D_A1_Y.resize(0);
       D_A1_MeasureErrors.resize(0);
       I_A1_OrigPos.resize(0);
       P_I_A1_Rejected->resize(0);
+      #ifdef __DEBUG_POLYFIT__
+        cout << "pfs::drp::stella::math::CurveFitting::PolyFit: D_LReject_In = " << D_LReject_In << ", D_UReject_In = " << D_UReject_In << endl;
+      #endif
       for (size_t i_pos=0; i_pos < D_A1_Y_In.getShape()[0]; i_pos++){
         D_Dev = D_A1_Y_In[i_pos] - D_A1_PolyRes[i_pos];
+        #ifdef __DEBUG_POLYFIT__
+          cout << "pfs::drp::stella::math::CurveFitting::PolyFit: i_pos = " << i_pos << ": D_Dev = " << D_Dev << ", D_SDev = " << D_SDev << endl;
+        #endif
         if (((D_Dev < 0) && (D_Dev > (D_LReject_In * D_SDev))) || 
             ((D_Dev >= 0) && (D_Dev < (D_UReject_In * D_SDev)))){
-          D_A1_X.push_back(xNew[i_pos]);
+          D_A1_X.push_back(D_A1_X_In[i_pos]);
```

Comment:
The fix is simple and unarguable. The code is long and hard to read, so I've not fought through all the implications, but this certainly looks right.

But: this seems like something which should be easily caught in a unit test. I wonder why we weren't hitting it before? Can you add a test to demonstrate that this code now does the right thing?

Updated later to add: I see there is now a test for PolyFit: great! Can you confirm that it would have caught this problem?